File: CSC205-Java/Module09-Stacks-and-Queues/debug02/CircularQueue.java
// A FIFO queue of integers stored in a circular (ring) array.
public class CircularQueue {
    private int[] data;
    private int front;   // index of the oldest item
    private int rear;    // index where the next item will be stored
    private int size;

    public CircularQueue(int capacity) {
        data = new int[capacity];
        front = 0;
        rear = 0;
        size = 0;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public boolean isFull() {
        return size == data.length;
    }

    public void enqueue(int value) {
        if (isFull()) {
            throw new RuntimeException("queue is full");
        }
        data[rear] = value;
        rear = rear + 1;
        size++;
    }

    public int dequeue() {
        if (isEmpty()) {
            throw new RuntimeException("queue is empty");
        }
        int value = data[front];
        front = (front + 1) % data.length;
        size--;
        return value;
    }

    public String toString() {
        StringBuilder sb = new StringBuilder("[");
        for (int i = 0; i < size; i++) {
            if (i > 0) {
                sb.append(", ");
            }
            sb.append(data[(front + i) % data.length]);
        }
        return sb.append("]").toString();
    }
}

File: CSC205-Cpp/_solutions/Module09-Stacks-and-Queues/debug02/circular_queue.cpp
// circular_queue.cpp
// A queue of print jobs stored in a fixed-size circular array. The driver
// fills the queue, prints a couple of jobs, adds more, and drains it.

#include <iostream>
using namespace std;

class CircularQueue {
private:
    int* items;
    int capacity;
    int front;      // index of the oldest item
    int rear;       // index where the next item will be stored
    int count;      // number of items currently in the queue

public:
    CircularQueue(int cap) : capacity(cap), front(0), rear(0), count(0) {
        items = new int[capacity];
    }

    ~CircularQueue() { delete[] items; }

    bool isEmpty() const { return count == 0; }
    bool isFull() const { return count == capacity; }
    int size() const { return count; }

    // Adds value at the rear. Returns false if the queue is full.
    bool enqueue(int value) {
        if (isFull()) {
            return false;
        }
        items[rear] = value;
        rear = (rear + 1) % capacity;
        count++;
        return true;
    }

    // Removes the oldest item into value. Returns false if the queue is empty.
    bool dequeue(int& value) {
        if (isEmpty()) {
            return false;
        }
        value = items[front];
        front = (front + 1) % capacity;
        count--;
        return true;
    }
};

void addJob(CircularQueue& q, int id) {
    cout << "add job " << id << ": " << (q.enqueue(id) ? "queued" : "rejected (full)") << endl;
}

void printJobs(CircularQueue& q, int howMany) {
    for (int i = 0; i < howMany; i++) {
        int job;
        if (q.dequeue(job)) {
            cout << "  printed job " << job << endl;
        } else {
            cout << "  nothing to print" << endl;
        }
    }
}

int main() {
    CircularQueue jobs(5);
    for (int id = 101; id <= 105; id++) {
        addJob(jobs, id);
    }
    cout << "size: " << jobs.size() << endl;

    cout << "Printing two jobs:" << endl;
    printJobs(jobs, 2);

    addJob(jobs, 106);
    addJob(jobs, 107);
    cout << "size: " << jobs.size() << endl;

    cout << "Printing everything:" << endl;
    printJobs(jobs, 6);
    return 0;
}
